core/crypto/poseidon_hash.go

```diff
@@ -0,0 +1,197 @@
+package crypto
+
+import (
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+const (
+	fullRounds    = 8
+	partialRounds = 83
+)
+
+func addRoundKeys(state []felt.Felt, round int) {
+	state[0].Add(&state[0], &roundKeys[round][0])
+	state[1].Add(&state[1], &roundKeys[round][1])
+	state[2].Add(&state[2], &roundKeys[round][2])
+}
+
+func subWords(state []felt.Felt, full bool) {
+	squared := felt.Felt{}
+	state[2].Mul(&state[2], squared.Mul(&state[2], &state[2]))
+	if full {
+		state[0].Mul(&state[0], squared.Mul(&state[0], &state[0]))
+		state[1].Mul(&state[1], squared.Mul(&state[1], &state[1]))
+	}
+}
+
+// M = ((3,1,1), (1,-1,1), (1,1,-2))
+// state = M * state
+func mixLayer(state []felt.Felt) {
+	stateSum := new(felt.Felt).Add(&state[0], &state[1])
+	stateSum.Add(stateSum, &state[2])
+	state[0].Double(&state[0]).Add(&state[0], stateSum)                               // stateSum + state[0] * 2
+	state[1].Sub(stateSum, state[1].Double(&state[1]))                                // stateSum - state[1] * 2
+	state[2].Sub(stateSum, state[2].Add(&state[2], new(felt.Felt).Double(&state[2]))) // stateSum - state[2] * 3
+}
+
+func round(state []felt.Felt, full bool, index int) {
+	addRoundKeys(state, index)
+	subWords(state, full)
+	mixLayer(state)
+}
+
+func permutate(state []felt.Felt) {
+	totalRounds := fullRounds + partialRounds
+	for i := 0; i < totalRounds; i++ {
+		full := (i < fullRounds/2) || (totalRounds-i <= fullRounds/2)
+		round(state, full, i)
+	}
+}
+
+var two = new(felt.Felt).SetUint64(2)
+
+// https://github.com/starkware-libs/cairo-lang/blob/12ca9e91bbdc8a423c63280949c7e34382792067/src/starkware/cairo/common/builtin_poseidon/poseidon.cairo#L5
+func Poseidon(x *felt.Felt, y *felt.Felt) *felt.Felt {
+	state := []felt.Felt{*x, *y, *two}
+	permutate(state)
+	return new(felt.Felt).Set(&state[0])
+}
+
+// PoseidonArray calculates Poseidon hash over elems
+// If len(elems) is odd, pads with [1]
+// If len(elems) is even, pads with [1, 0]
+//
+// https://github.com/starkware-libs/cairo-lang/blob/12ca9e91bbdc8a423c63280949c7e34382792067/src/starkware/cairo/common/builtin_poseidon/poseidon.cairo#L28
+func PoseidonArray(elems ...*felt.Felt) *felt.Felt {
+	state := []felt.Felt{{}, {}, {}}
+
+	for i := 0; i < len(elems)/2; i++ {
+		state[0].Add(&state[0], elems[2*i])
+		state[1].Add(&state[1], elems[2*i+1])
+		permutate(state)
+	}
+
+	rem := len(elems) % 2
+	if rem == 1 {
+		state[0].Add(&state[0], elems[len(elems)-1])
+	}
+	state[rem].Add(&state[rem], new(felt.Felt).SetUint64(1))
+	permutate(state)
+
+	return new(felt.Felt).Set(&state[0])
+}
```

Comment:
This does not work correctly if the length of the array is 1 or 2. Just like in our pedersen hash function, I think this should also be able to produce the correct hash whether the length is 0, 1, or 2.

Reply:
Poseidon2 and PoseidonArray is 2 different functions. They are not supposed to yield the same results.

see https://docs.starknet.io/documentation/architecture_and_concepts/Hashing/hash-functions/

Reply:
Can we have Poseidon1 implemented as well then?

Reply:
Of course, but it is not really used anywhere at this point.

Reply:
Was able to do some comparison test for this and yes, it works perfectly. Great work, boss

Reply:
We should try generating tests by running random values through Cairo Lang's Python implementation